### PR TITLE
feat: Library for sorted set validations

### PR DESCRIPTION
**Pull request by 0xsuryansh**
### Motivation
We have multiple validation use-cases on-chain which requires the inputs to be a set, sorted-set or we need to do subset checks.
Adding a library for these validations

### Resolves Dependencies  
Does this work support other open PRs? 
https://github.com/smartcontractkit/ccip/pull/1129

### CODE REVIEW

File: contracts/src/v0.8/shared/util/SortedSetValidationUtil.sol [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  /// @dev Error to be thrown when the first array is not a subset of the second array.
+  error NotASubset(bytes32[] subset, bytes32[] superset);
+
+  /// @notice Checks if `subset` is a valid and unique subset of `superset`.
+  /// NOTE: Empty set is not considered a valid subset of superset for our use case.
+  /// @dev Both arrays must be valid sets (unique, sorted in ascending order) and `subset` must be entirely contained within `superset`.
+  /// @param subset The array of bytes32 to validate as a subset.
+  /// @param superset The array of bytes32 in which subset is checked against.
+  /// @custom:revert EmptySet If either `subset` or `superset` is empty.
+  /// @custom:revert NotASubset If `subset` is not a subset of `superset`.
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/util/SortedSetValidationUtil.sol:21]: suggestion: we should add a comment that it also verifies that both `subset` and `superset` are sorted

File: contracts/src/v0.8/shared/test/util/SortedSetValidationUtil.t.sol [hunk 2]
```diff
@@ -0,0 +1,112 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+import {BaseTest} from "../BaseTest.t.sol";
+import {SortedSetValidationUtil} from "../../../shared/util/SortedSetValidationUtil.sol";
+
+contract SortedSetValidationUtilTest is BaseTest {
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/test/util/SortedSetValidationUtil.t.sol]: nit: could rename it to `SortedSetValidationUtilBaseTest`

File: contracts/src/v0.8/shared/test/util/SortedSetValidationUtil.t.sol [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    _makeInvalidSubset(subset, superset, 1);
+
+    vm.expectRevert(abi.encodeWithSelector(SortedSetValidationUtil.NotASubset.selector, subset, superset));
+    SortedSetValidationUtil._checkIsValidUniqueSubset(subset, superset);
+  }
+
+  function test__checkIsValidUniqueSubset_NotASortedSet_Reverts() public {
+    (bytes32[] memory subset, bytes32[] memory superset) = _createSets(3, 5);
+    _convertArrayToSortedSet(superset, OFFSET);
+    _convertArrayToUnsortedSet(subset, OFFSET, 1, 2);
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/test/util/SortedSetValidationUtil.t.sol]: suggestion: we should add a test-case where the `superset` is unsorted and the subset is sorted

File: contracts/src/v0.8/shared/test/util/SortedSetValidationUtil.t.sol [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+  function test__checkIsValidUniqueSubset_EmptySuperset_Reverts() public {
+    (bytes32[] memory subset, bytes32[] memory superset) = _createSets(3, 0);
+    _convertArrayToSortedSet(subset, OFFSET);
+
+    vm.expectRevert(abi.encodeWithSelector(SortedSetValidationUtil.EmptySet.selector));
+    SortedSetValidationUtil._checkIsValidUniqueSubset(subset, superset);
+  }
+
+  function test__checkIsValidUniqueSubset_NotASubset_Reverts() public {
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/test/util/SortedSetValidationUtil.t.sol:79]: suggestion: we should add more test cases:
* Subset.length > Superset.length (revert)
* subset == superset (no revert)
* Subset.length = 1 && superset.length > 1
* subset.length = 1 && superset.length = 1 (equal)
* subset.length = 1 && superset.length = 1 (not equal)

File: contracts/src/v0.8/shared/test/util/SortedSetValidationUtil.t.sol [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  // Successes.
+
+  function test__checkIsValidUniqueSubset_ValidSubset_Success() public {
+    (bytes32[] memory subset, bytes32[] memory superset) = _createSets(3, 5);
+    _convertArrayToSortedSet(superset, OFFSET);
+    _convertArrayToSubset(subset, superset);
+
+    SortedSetValidationUtil._checkIsValidUniqueSubset(subset, superset);
+  }
+
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/test/util/SortedSetValidationUtil.t.sol:60]: suggestion: should we create a fuzz test for the util library as well? Could set it to a very low # of runs

File: contracts/src/v0.8/shared/test/util/SortedSetValidationUtil.t.sol [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+  function test__checkIsValidUniqueSubset_NotASortedSet_Reverts() public {
+    (bytes32[] memory subset, bytes32[] memory superset) = _createSets(3, 5);
+    _convertArrayToSortedSet(superset, OFFSET);
+    _convertArrayToUnsortedSet(subset, OFFSET, 1, 2);
+
+    vm.expectRevert(abi.encodeWithSelector(SortedSetValidationUtil.NotASortedSet.selector, subset));
+    SortedSetValidationUtil._checkIsValidUniqueSubset(subset, superset);
+  }
+
+  function test__checkIsValidUniqueSubset_HasDuplicates_Reverts() public {
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/test/util/SortedSetValidationUtil.t.sol:107]: suggestion: we should also add a test-case where the `superset` has duplicates instead of the subset

File: contracts/src/v0.8/shared/util/SortedSetValidationUtil.sol (hunk 1 above)
Comment (0xsuryansh) [reply contracts/src/v0.8/shared/util/SortedSetValidationUtil.sol:21]: I included that in the contract's docstring, Line 6
will add this for this function as well

File: contracts/src/v0.8/shared/test/util/SortedSetValidationUtil.t.sol (hunk 4 above)
Comment (0xsuryansh) [reply contracts/src/v0.8/shared/test/util/SortedSetValidationUtil.t.sol:79]: https://github.com/smartcontractkit/chainlink/pull/13774/commits/4578231b9ca5319a6209b2d32762c1347a8cb72f

Added all the suggested tests here

File: contracts/src/v0.8/shared/test/util/SortedSetValidationUtil.t.sol (hunk 5 above)
Comment (0xsuryansh) [reply contracts/src/v0.8/shared/test/util/SortedSetValidationUtil.t.sol:60]: Decided to add fuzz tests for CCIPConfig

File: contracts/src/v0.8/shared/test/util/SortedSetValidationUtil.t.sol (hunk 2 above)
Comment (0xsuryansh) [reply contracts/src/v0.8/shared/test/util/SortedSetValidationUtil.t.sol]: Done

File: contracts/src/v0.8/shared/test/util/SortedSetValidationUtil.t.sol [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+  function test__checkIsValidUniqueSubset_NotASubset_Reverts() public {
+    (bytes32[] memory subset, bytes32[] memory superset) = _createSets(3, 5);
+    _convertArrayToSortedSet(superset, OFFSET);
+    _makeInvalidSubset(subset, superset, 1);
+
+    vm.expectRevert(abi.encodeWithSelector(SortedSetValidationUtil.NotASubset.selector, subset, superset));
+    SortedSetValidationUtil._checkIsValidUniqueSubset(subset, superset);
+  }
+
+  function test__checkIsValidUniqueSubset_NotASortedSet_Reverts() public {
```
Comment (elatoskinas) [inline contracts/src/v0.8/shared/test/util/SortedSetValidationUtil.t.sol]: nit: for consistency we should use the same naming pattern `NotASortedSubset` / `NotASortedSuperset` or `UnsortedSuperSet` / `UnsortedSubset`

**Review by elatoskinas — DISMISSED** [review]

File: contracts/src/v0.8/shared/test/util/SortedSetValidationUtil.t.sol (hunk 7 above)
Comment (0xsuryansh) [reply contracts/src/v0.8/shared/test/util/SortedSetValidationUtil.t.sol]: https://github.com/smartcontractkit/chainlink/pull/13774/commits/2146c8516c1c097b22e1a1e6a0f8400529f0b0f8

**Review by makramkd — DISMISSED** [review]

**Review by elatoskinas — DISMISSED** [review]

**Review by elatoskinas — DISMISSED** [review]

**Review by makramkd — APPROVED** [review]

**Review by elatoskinas — APPROVED** [review]
